Include/descrobject.h

```diff
@@ -93,7 +93,7 @@ PyAPI_FUNC(PyObject *) PyDescr_NewGetSet(PyTypeObject *,
 #ifndef Py_LIMITED_API
 PyAPI_FUNC(PyObject *) PyDescr_NewWrapper(PyTypeObject *,
                                                 struct wrapperbase *, void *);
-#define PyDescr_IsData(d) (Py_TYPE(d)->tp_descr_set != NULL)
+PyAPI_FUNC(int) PyDescr_IsData(PyObject *);
```

Comment:
Would it make sense to add it the limited C API? The function is documented at:
https://docs.python.org/dev/c-api/descriptor.html?highlight=pydescr_isdata#c.PyDescr_IsData

It seems like most of the Include/descrobject.h (including PyDescrObject structure!) is not excluded from the limited C API.

If you add it, you should write something like:

```
#if !defined(Py_LIMITED_API) || Py_LIMITED_API+0 >= 0x03030000
```

To only expose it since Python 3.10.

Reply:
> Would it make sense to add it the limited C API? The function is documented at:
> https://docs.python.org/dev/c-api/descriptor.html?highlight=pydescr_isdata#c.PyDescr_IsData

My initial thought is, yes that would make sense. I'd like to hear @encukou's opinion before deciding, though.

> It seems like most of the Include/descrobject.h (including PyDescrObject structure!) is not excluded from the limited C API.

Yes, so I've seen! (Quoting from PEP 384: _"The structure of type objects is not available to applications"_)

Also, a quick code search reveals that both `PyDescr_NAME` and `PyDescr_TYPE` are used as l-values in external projects, so those won't be straight-forward.
 
> If you add it, you should write something like:
> 
> ```
> #if !defined(Py_LIMITED_API) || Py_LIMITED_API+0 >= 0x03030000
> ```
> 
> To only expose it since Python 3.10.

Thanks!

Reply:
The `Py_LIMITED_API` define is not authoritative. IMO, PEP 384 is – at least for things it mentions explicitly.

`PyDescr_IsData` itself looks like a reaasonable addition to the limited API, but please document it more thoroughly: what does it mean that descr “must be a descriptor object”? How would you check? What happens if it isn't? (“there is no error checking” sounds like too much undefined behavior to me)
It seems like a  `PyDescr_Check` would be a good addition to go along with `PyDescr_IsData`.

Or maybe it would be enough to instead use `PyType_GetSlot` on `Py_tp_descr_get` & `Py_tp_descr_set`?

Reply:
> [...] what does it mean that descr “must be a descriptor object”? How would you check? What happens if it isn't? (“there is no error checking” sounds like too much undefined behavior to me)

I think opening a separate ticket on bpo for improving all of https://docs.python.org/3/c-api/descriptor.html would be a good idea.

> It seems like a `PyDescr_Check` would be a good addition to go along with `PyDescr_IsData`.

That sounds out of scope for this PR. Also, `PyDescr_Check` does not exist. If it did, what should `PyDescr_IsData` return if _descr_ wasn't a descriptor object?

> Or maybe it would be enough to instead use `PyType_GetSlot` on `Py_tp_descr_get` & `Py_tp_descr_set`?

Maybe, but that will alter its behaviour slightly, no? I'd prefer the current implementation.

Reply:
@vstinner: Looking over `Include/descrobject.h` again, I see that `PyDescrObject and most other structs are actually wrapped in `#ifndef Py_LIMITED_API`, so they're excluded. `PyDescr_NAME` and `PyDescr_TYPE` are also excluded from the limited API.

Reply:
> ```
> #if !defined(Py_LIMITED_API) || Py_LIMITED_API+0 >= 0x03030000
> ```
> 
> To only expose it since Python 3.10.

`0x030A0000`, not `0x03030000`, I presume.

Reply:
> what does it mean that descr “must be a descriptor object”?

It's an object created by a PyDescr_NewXXX() function documented just before, I guess.

> How would you check?

You cannot.

> What happens if it isn't? (“there is no error checking” sounds like too much undefined behavior to me)

It does crash.

> It seems like a PyDescr_Check would be a good addition to go along with PyDescr_IsData.

Honestly, I never understood exactly what are descriptors :-p I'm not sure if people are supposed to call PyDescr_IsData() outside CPython internals. But it's a public function. Was it exposed on purpose or by mistake? :-D

Maybe it's better to keep it outside the limited C API until someone asks to add it?

Reply:
> Honestly, I never understood exactly what are descriptors :-p I'm not sure if people are supposed to call PyDescr_IsData() outside CPython internals. But it's a public function. Was it exposed on purpose or by mistake? :-D
> 
> Maybe it's better to keep it outside the limited C API until someone asks to add it?

I guess so :) [searchcode.com](https://searchcode.com/?q=pydescr_isdata) yields 144 hits, but most of them seems to be CPython mirrors.

Reply:
I suggest reverting 7d71a1a5c26a8289d70e599b4dd2c6763340290a

Reply:
Sorry, please revert your commit. I am no longer sure that it's a good idea to add it the limited C API.

Reply:
No problem; done! I consider this conversation resolved :)